### ISSUE TITLE
DevTools sometimes fails to precompile (maybe when an update is available)

### ISSUE THREAD

**Issue by zanderso** [flutter/flutter#82142]
https://ci.chromium.org/p/flutter/builders/prod/Linux%20complex_layout_scroll_perf__devtools_memory

P0 because this is closing the tree a lot.

This is exactly like https://github.com/flutter/flutter/issues/81513 again but with 2.1.1 instead of 2.0.0. Is there another new version available? /cc @jacob314 @kenzieschmoll @jonahwilliams 

Very strange that sometimes it succeeds and sometimes it fails with error message:
```
[complex_layout_scroll_perf__devtools_memory] [STDOUT] Executing: /opt/s/w/ir/x/w/recipe_cleanup/tmpl7tXfX/flutter sdk/bin/cache/dart-sdk/bin/pub global activate devtools 2.1.1 in /opt/s/w/ir/x/w/recipe_cleanup/tmpl7tXfX/flutter sdk/dev/benchmarks/complex_layout
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: Resolving dependencies...
[complex_layout_scroll_perf__devtools_memory] [STDOUT] run stdout: [ +158 ms] Activating Dart DevTools...
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + args 2.1.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + async 2.6.1
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + browser_launcher 0.1.10 (1.0.0 available)
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + charcode 1.2.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + clock 1.1.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + collection 1.15.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + convert 3.0.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + devtools 2.1.1 (2.2.2 available)
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + devtools_server 2.1.1 (2.2.2 available)
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + devtools_shared 2.1.1 (2.2.2 available)
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + http 0.13.3
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + http_multi_server 3.0.1
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + http_parser 4.0.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + intl 0.17.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + logging 1.0.1
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + meta 1.3.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + mime 1.0.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + path 1.8.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + pedantic 1.11.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + shelf 1.1.2
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + shelf_proxy 1.0.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + shelf_static 1.0.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + source_span 1.8.1
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + sse 3.8.3
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + stack_trace 1.10.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + stream_channel 2.1.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + string_scanner 1.1.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + term_glyph 1.2.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + typed_data 1.3.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + usage 4.0.2
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + vm_service 6.2.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: + webkit_inspection_protocol 1.0.0
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: Downloading devtools 2.1.1...
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: Downloading devtools_shared 2.1.1...
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: Downloading devtools_server 2.1.1...
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: Downloading async 2.6.1...
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: Downloading http 0.13.3...
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: Downloading shelf 1.1.2...
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: Downloading browser_launcher 0.1.10...
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stdout: Precompiling executables...
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr: Failed to precompile devtools:devtools:
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr: Error: Cannot run with sound null safety, because the following dependencies
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr: don't support null safety:
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr: 
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr:  - package:devtools_server
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr:  - package:browser_launcher
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr:  - package:devtools_shared
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr:  - package:sse
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr: 
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr: For solutions, see https://dart.dev/go/unsound-null-safety
```

**Comment by jonahwilliams** [comment]
I don't understand why releasing a new version was cause the old version to fail compilation - unless we were breaking some invariants that pub depends on for the precompiled snapshots. I can take a look at this on monday to see if we could just use the devtools that the tool is serving? (Unless that is also still broken)

**Comment by jacob314** [comment]
Would be great to find the real issue.
https://github.com/flutter/flutter/pull/81869
should unblock.

**Comment by zanderso** [comment]
This is affecting other tests as well: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20large_image_changer_perf_android/1315/overview

**Comment by zanderso** [comment]
PRs appear to be blocked on https://github.com/flutter/flutter/issues/82225

**Comment by jacob314** [comment]
I believe the error is because for some reson the devtools build in this test is panicking that these packages are not migrated to null safety and it only panics if the version changes. 3 of the 4 are migrated and @kenzieschmoll is migrating the last packge.
I wish we could understand why this error is showing up on the bots even though we can't reproduce but fixing the packages should at least make the error go awy or clarify why it is occurring.
complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr:  - package:devtools_server
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr:  - package:browser_launcher
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr:  - package:devtools_shared
[complex_layout_scroll_perf__devtools_memory] [STDOUT] stderr:  - package:sse

**Comment by christopherfujino** [comment]
I landed https://github.com/flutter/flutter/pull/82491 to deactivate devtools before trying to install it--not sure if it solves this flaky compilation error.

**Comment by zanderso** [comment]
It looks like following https://github.com/flutter/flutter/pull/82491 this is failing a different way?
```
[large_image_changer_perf_android] [STDOUT] run stdout: [ +192 ms] Activating Dart DevTools...
[large_image_changer_perf_android] [STDOUT] stdout: Resolving dependencies...
[large_image_changer_perf_android] [STDOUT] stdout: + args 2.1.0
[large_image_changer_perf_android] [STDOUT] stdout: + async 2.6.1
[large_image_changer_perf_android] [STDOUT] stdout: + browser_launcher 0.1.10 (1.0.0 available)
[large_image_changer_perf_android] [STDOUT] stdout: + charcode 1.2.0
[large_image_changer_perf_android] [STDOUT] stdout: + clock 1.1.0
[large_image_changer_perf_android] [STDOUT] stdout: + collection 1.15.0
[large_image_changer_perf_android] [STDOUT] stdout: + convert 3.0.0
[large_image_changer_perf_android] [STDOUT] stdout: + devtools 2.2.3
[large_image_changer_perf_android] [STDOUT] stdout: + devtools_server 2.2.3
[large_image_changer_perf_android] [STDOUT] stdout: + devtools_shared 2.2.3 (2.3.0 available)
[large_image_changer_perf_android] [STDOUT] stdout: + http 0.13.3
[large_image_changer_perf_android] [STDOUT] stdout: + http_multi_server 3.0.1
[large_image_changer_perf_android] [STDOUT] stdout: + http_parser 4.0.0
[large_image_changer_perf_android] [STDOUT] stdout: + intl 0.17.0
[large_image_changer_perf_android] [STDOUT] stdout: + logging 1.0.1
[large_image_changer_perf_android] [STDOUT] stdout: + meta 1.3.0
[large_image_changer_perf_android] [STDOUT] stdout: + mime 1.0.0
[large_image_changer_perf_android] [STDOUT] stdout: + path 1.8.0
[large_image_changer_perf_android] [STDOUT] stdout: + pedantic 1.11.0
[large_image_changer_perf_android] [STDOUT] stdout: + shelf 1.1.4
[large_image_changer_perf_android] [STDOUT] stdout: + shelf_proxy 1.0.0
[large_image_changer_perf_android] [STDOUT] stdout: + shelf_static 1.0.0
[large_image_changer_perf_android] [STDOUT] stdout: + source_span 1.8.1
[large_image_changer_perf_android] [STDOUT] stdout: + sse 3.8.3 (4.0.0 available)
[large_image_changer_perf_android] [STDOUT] stdout: + stack_trace 1.10.0
[large_image_changer_perf_android] [STDOUT] stdout: + stream_channel 2.1.0
[large_image_changer_perf_android] [STDOUT] stdout: + string_scanner 1.1.0
[large_image_changer_perf_android] [STDOUT] stdout: + term_glyph 1.2.0
[large_image_changer_perf_android] [STDOUT] stdout: + typed_data 1.3.0
[large_image_changer_perf_android] [STDOUT] stdout: + usage 4.0.2
[large_image_changer_perf_android] [STDOUT] stdout: + vm_service 6.2.0 (7.0.0 available)
[large_image_changer_perf_android] [STDOUT] stdout: + webkit_inspection_protocol 1.0.0
[large_image_changer_perf_android] [STDOUT] stdout: Downloading devtools 2.2.3...
[large_image_changer_perf_android] [STDOUT] stdout: Downloading devtools_shared 2.2.3...
[large_image_changer_perf_android] [STDOUT] stdout: Downloading devtools_server 2.2.3...
[large_image_changer_perf_android] [STDOUT] stdout: Downloading browser_launcher 0.1.10...
[large_image_changer_perf_android] [STDOUT] stderr: Directory listing failed, path = '/opt/s/w/ir/x/w/.pub-cache/hosted/pub.dartlang.org/devtools-2.2.3/bin/' (OS Error: No such file or directory, errno = 2)
[large_image_changer_perf_android] [STDOUT] "/opt/s/w/ir/x/w/recipe_cleanup/tmpV0LtUf/flutter sdk/bin/cache/dart-sdk/bin/pub" exit code: 66
```

**Comment by christopherfujino** [comment]
> [large_image_changer_perf_android] [STDOUT] stderr: Directory listing failed, path = '/opt/s/w/ir/x/w/.pub-cache/hosted/pub.dartlang.org/devtools-2.2.3/bin/' (OS Error: No such file or directory, errno = 2)
> [large_image_changer_perf_android] [STDOUT] "/opt/s/w/ir/x/w/recipe_cleanup/tmpV0LtUf/flutter sdk/bin/cache/dart-sdk/bin/pub" exit code: 66
> ```

Full logs available here: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8847021809092651408/+/u/run_large_image_changer_perf_android/stdout

On this build it tried to deactiave devtools, however that failed with `No active package devtools`, which should be fine. The script then attempted `pub global activate devtools 2.2.3` and failed as above. I'm not sure what went wrong.

**Comment by christopherfujino** [comment]
Merged https://github.com/flutter/flutter/commit/d6d3f0e0a82b8c663a82978f0c22c22d50637112 to add more logs.

**Comment by Hixie** [comment]
Any news from the logs?

**Comment by jonahwilliams** [comment]
This seems like it may have been fixed by https://github.com/flutter/flutter/pull/82739 

My guess is that there might have been a race condition between the devtools activated by flutter run and by the test script, since they run at about the same time

**Comment by jonahwilliams** [comment]
This test was also more complicated than it needed to be due to the separate `run` / `drive` /`uninstall` commands instead of a single drive command.

**Comment by christopherfujino** [comment]
> This seems like it may have been fixed by #82739

Yeah, this has not failed since I added verbose logs, and Jonah's fix landed a few commits later. I will optimistically close this issue (and mark the test un-flaky), feel free to re-open this issue if the compilation error resurfaces.
